packages/flutter/lib/src/widgets/lazy_block.dart

```diff
@@ -0,0 +1,449 @@
+    snapOffsetCallback: snapOffsetCallback
+  );
+
+  final LazyBlockDelegate delegate;
+
+  @override
+  ScrollableState<LazyBlock> createState() => new _LazyBlockState();
+}
+
+class _LazyBlockState extends ScrollableState<LazyBlock> {
+  @override
+  BoundedBehavior createScrollBehavior() => new OverscrollWhenScrollableBehavior();
+
+  @override
+  BoundedBehavior get scrollBehavior => super.scrollBehavior;
+
+  void _handleExtentsChanged(double contentExtent, double containerExtent, double minScrollOffset) {
+    setState(() {
+      didUpdateScrollBehavior(scrollBehavior.updateExtents(
+        contentExtent: contentExtent,
+        containerExtent: containerExtent,
+        minScrollOffset: minScrollOffset,
+        scrollOffset: scrollOffset
+      ));
+    });
+  }
+
+  @override
+  Widget buildContent(BuildContext context) {
+    return new LazyBlockViewport(
+      startOffset: scrollOffset,
+      direction: config.scrollDirection,
+      onExtentsChanged: _handleExtentsChanged,
+      delegate: config.delegate
+    );
+  }
+}
+
+typedef void LazyBlockExtentsChangedCallback(double contentExtent, double containerExtent, double minScrollOffset);
+
+class LazyBlockViewport extends RenderObjectWidget {
+  LazyBlockViewport({
+    Key key,
+    this.startOffset,
+    this.direction,
+    this.onExtentsChanged,
+    this.delegate
+  }) : super(key: key);
+
+  final double startOffset;
+  final Axis direction;
+  final LazyBlockExtentsChangedCallback onExtentsChanged;
+  final LazyBlockDelegate delegate;
+
+  @override
+  _LazyBlockElement createElement() => new _LazyBlockElement(this);
+
+  @override
+  _RenderLazyBlock createRenderObject(BuildContext context) => new _RenderLazyBlock();
+}
+
+class _LazyBlockParentData extends ContainerBoxParentDataMixin<RenderBox> { }
+
+class _RenderLazyBlock extends RenderVirtualViewport<_LazyBlockParentData> {
+  _RenderLazyBlock({
+    Offset paintOffset: Offset.zero,
+    Axis mainAxis: Axis.vertical,
+    LayoutCallback callback
+  }) : super(
+    paintOffset: paintOffset,
+    mainAxis: mainAxis,
+    callback: callback
+  );
+
+  @override
+  void setupParentData(RenderBox child) {
+    if (child.parentData is! _LazyBlockParentData)
+      child.parentData = new _LazyBlockParentData();
+  }
+
+  @override
+  double getMinIntrinsicWidth(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsNormalized);
+    return constraints.constrainWidth(0.0);
+  }
+
+  @override
+  double getMaxIntrinsicWidth(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsNormalized);
+    return constraints.constrainWidth(0.0);
+  }
+
+  @override
+  double getMinIntrinsicHeight(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsNormalized);
+    return constraints.constrainHeight(0.0);
+  }
+
+  @override
+  double getMaxIntrinsicHeight(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsNormalized);
+    return constraints.constrainHeight(0.0);
+  }
+
```

Comment:
the intrinsic dimension in the cross-axis should not be queriable. There's code in mixed viewport to report this that you can clone (it's a little tricky because it has to not trigger on the sanity checks).